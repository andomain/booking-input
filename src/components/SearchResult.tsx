import React from 'react';

const SearchResult = () => {
  return <li></li>;
};

export default SearchResult;
