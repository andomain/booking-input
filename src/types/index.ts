export enum PLACE_TYPE {
  AIRPORT = 'A',
  DISTRICT = 'D',
  CITY = 'C',
  STATION = 'T',
}

// Subset of response data shared by all response types
type BaseResponse = {
  country: string;
  placeKey: string;
  name: string;
  region: string;
};

// Extend base types with data specific to airports etc.
type AIRPORT = BaseResponse & {
  placeType: PLACE_TYPE.AIRPORT;
  iata: string;
  city?: string;
};

type CITY = BaseResponse & {
  placeType: PLACE_TYPE.CITY;
};
type DISTRICT = BaseResponse & {
  placeType: PLACE_TYPE.DISTRICT;
};
type STATION = BaseResponse & {
  placeType: PLACE_TYPE.STATION;
};

export type SearchResponse = AIRPORT | CITY | STATION | DISTRICT;
