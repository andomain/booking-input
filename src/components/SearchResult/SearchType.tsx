import React, { FC } from 'react';
import classnames from 'classnames';

import { PLACE_TYPE } from '../../types';

interface SearchTypeProps {
  type: PLACE_TYPE;
  className?: string;
}

const getLabel = (type: PLACE_TYPE): string => {
  switch (type) {
    case PLACE_TYPE.AIRPORT:
      return 'Airport';
    case PLACE_TYPE.STATION:
      return 'Station';
    case PLACE_TYPE.DISTRICT:
      return 'District';
    case PLACE_TYPE.CITY:
      return 'City';
    default:
      return 'Unknown';
  }
};

const SearchType: FC<SearchTypeProps> = ({ className = '', type }) => {
  const classes = classnames('SearchType__Pill', {
    'SearchType__Pill--city': type === PLACE_TYPE.CITY,
    'SearchType__Pill--airport': type === PLACE_TYPE.AIRPORT,
    'SearchType__Pill--district': type === PLACE_TYPE.DISTRICT,
    'SearchType__Pill--station': type === PLACE_TYPE.STATION,
  });

  return (
    <div className={`SearchType ${className}`}>
      <span className={classes}>{getLabel(type)}</span>
    </div>
  );
};

export default SearchType;
