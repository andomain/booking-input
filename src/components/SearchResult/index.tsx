import React, { FC } from 'react';

import SearchType from './SearchType';

import { SearchResponse, PLACE_TYPE } from '../../types';

interface SearchResultProps {
  data: SearchResponse;
  onSelect: (name: string, data: SearchResponse) => void;
}

const buildName = (data: SearchResponse): string => {
  if (data.placeType === PLACE_TYPE.AIRPORT && data.iata) {
    return `${data.name} (${data.iata})`;
  }
  return data.name;
};

const buildLocation = (data: SearchResponse): string => {
  if (data.placeType === PLACE_TYPE.AIRPORT) {
    if (data.city) {
      return `${data.city}, ${data.country}`;
    }
    return data.country;
  }
  return `${data.region}, ${data.country}`;
};

const SearchResult: FC<SearchResultProps> = ({ data, onSelect }) => {
  const clickHandler = (displayName: string) => onSelect(name, data);

  const name = buildName(data);
  const location = buildLocation(data);

  return (
    <li className="SearchResult" onClick={() => clickHandler(name)}>
      <SearchType className="SearchResult__Type" type={data.placeType} />
      <div>
        <p className="SearchResult__Name">{name}</p>
        <p className="SearchResult__Sub">{location}</p>
      </div>
    </li>
  );
};

export default SearchResult;
