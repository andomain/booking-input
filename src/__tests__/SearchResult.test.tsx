import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import SearchResult from '../components/SearchResult';
import { PLACE_TYPE, SearchResponse } from '../types';

// Example city item
const CITY_DATA: SearchResponse = {
  country: 'United Kingdom',
  placeKey: 'abc',
  placeType: PLACE_TYPE.CITY,
  name: 'Leeds',
  region: 'West Yorkshire',
};

// Example airport data
const AIRPORT_DATA: SearchResponse = {
  country: 'United Kingdom',
  city: 'London',
  placeType: PLACE_TYPE.AIRPORT,
  placeKey: 'abc',
  iata: 'LGW',
  name: 'Gatwick Airport',
  region: 'Greater London',
};

const DISTRICT_DATA: SearchResponse = {
  country: 'United Kingdom',
  placeType: PLACE_TYPE.DISTRICT,
  placeKey: 'abc',
  name: 'Leeds City Centre',
  region: 'West Yorkshire',
};

const STATION_DATA: SearchResponse = {
  country: 'United Kingdom',
  city: 'Manchester',
  placeType: PLACE_TYPE.STATION,
  placeKey: 'abc',
  name: 'Manchester - Piccadilly Train Station',
  region: 'England',
};

describe('<SearchResult />', () => {
  let wrapper: ReactWrapper;

  const mockSelect = jest.fn();

  beforeEach(() => {
    wrapper = mount(<SearchResult onSelect={mockSelect} data={CITY_DATA} />);
  });

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('should render a <li />', () => {
    expect(wrapper.find('li').length).toBe(1);
  });

  it('should fire an onSelect handler when clicked', () => {
    wrapper.simulate('click');
    expect(mockSelect).toBeCalledTimes(1);
    expect(mockSelect).toBeCalledWith(CITY_DATA.name, CITY_DATA);
  });

  describe('Result name', () => {
    it('should display the name of the search result', () => {
      const name = wrapper.find('p.SearchResult__Name');
      expect(name.length).toEqual(1);
      expect(name.text()).toEqual(CITY_DATA.name);
    });

    it('should display the airport IATA code', () => {
      const airportWrapper = mount(
        <SearchResult onSelect={mockSelect} data={AIRPORT_DATA} />,
      );
      const name = airportWrapper.find('p.SearchResult__Name');
      expect(name.text()).toEqual(
        `${AIRPORT_DATA.name} (${AIRPORT_DATA.iata})`,
      );
    });
  });

  describe('Result types', () => {
    it('should display result type', () => {
      const type = wrapper.find('.SearchType__Pill');
      expect(type.length).toEqual(1);
    });

    it('should display city responses', () => {
      const type = wrapper.find('.SearchType__Pill');
      expect(type.text()).toEqual('City');
    });

    it('should display airport responses', () => {
      const airportWrapper = mount(
        <SearchResult onSelect={mockSelect} data={AIRPORT_DATA} />,
      );
      expect(airportWrapper.find('.SearchType__Pill').text()).toEqual(
        'Airport',
      );
    });

    it('should display district responses', () => {
      const districtWrapper = mount(
        <SearchResult onSelect={mockSelect} data={DISTRICT_DATA} />,
      );
      expect(districtWrapper.find('.SearchType__Pill').text()).toEqual(
        'District',
      );
    });

    it('should display station responses', () => {
      const stationWrapper = mount(
        <SearchResult onSelect={mockSelect} data={STATION_DATA} />,
      );
      expect(stationWrapper.find('.SearchType__Pill').text()).toEqual(
        'Station',
      );
    });

    it('should style based on response type', () => {
      const airportWrapper = mount(
        <SearchResult onSelect={mockSelect} data={AIRPORT_DATA} />,
      );
      const districtWrapper = mount(
        <SearchResult onSelect={mockSelect} data={DISTRICT_DATA} />,
      );
      const stationWrapper = mount(
        <SearchResult onSelect={mockSelect} data={STATION_DATA} />,
      );

      expect(wrapper.find('.SearchType__Pill').props().className).toContain(
        'SearchType__Pill--city',
      );
      expect(
        airportWrapper.find('.SearchType__Pill').props().className,
      ).toContain('SearchType__Pill--airport');
      expect(
        stationWrapper.find('.SearchType__Pill').props().className,
      ).toContain('SearchType__Pill--station');
      expect(
        districtWrapper.find('.SearchType__Pill').props().className,
      ).toContain('SearchType__Pill--district');
    });
  });

  describe('Location', () => {
    it('should display the region & country combined by default', () => {
      const location = wrapper.find('p.SearchResult__Sub');
      expect(location.length).toEqual(1);
      expect(location.text()).toEqual(
        `${CITY_DATA.region}, ${CITY_DATA.country}`,
      );
    });

    it('should display airport cities if present', () => {
      const airportWrapper = mount(
        <SearchResult onSelect={mockSelect} data={AIRPORT_DATA} />,
      );
      const location = airportWrapper.find('p.SearchResult__Sub');

      expect(location.text()).toEqual(
        `${AIRPORT_DATA.city}, ${AIRPORT_DATA.country}`,
      );
    });

    it('should only display airport country if no city present', () => {
      const TEST_DATA: SearchResponse = { ...AIRPORT_DATA, city: undefined };
      const airportWrapper = mount(
        <SearchResult onSelect={mockSelect} data={TEST_DATA} />,
      );
      const location = airportWrapper.find('p.SearchResult__Sub');

      expect(location.text()).toEqual(AIRPORT_DATA.country);
    });
  });
});
