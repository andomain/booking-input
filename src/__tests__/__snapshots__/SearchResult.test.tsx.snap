// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResult /> should render correctly 1`] = `
<SearchResult
  data={
    Object {
      "country": "United Kingdom",
      "name": "Leeds",
      "placeKey": "abc",
      "placeType": "C",
      "region": "West Yorkshire",
    }
  }
  onSelect={[MockFunction]}
>
  <li
    className="SearchResult"
    onClick={[Function]}
  >
    <SearchType
      className="SearchResult__Type"
      type="C"
    >
      <div
        className="SearchType SearchResult__Type"
      >
        <span
          className="SearchType__Pill SearchType__Pill--city"
        >
          City
        </span>
      </div>
    </SearchType>
    <div>
      <p
        className="SearchResult__Name"
      >
        Leeds
      </p>
      <p
        className="SearchResult__Sub"
      >
        West Yorkshire, United Kingdom
      </p>
    </div>
  </li>
</SearchResult>
`;
